### PR TITLE
Enable 'persist-redux' feature in every environment

### PR DESCRIPTION
**Pull request by rcanepa**
#### The problem

Currently, when the `siteSelection` controller can't find the selected site in the app's state (in `state.sites.items` to be precise), it sends a request to fetch the site's information. If the site has a broken connection, that request will fail. This will likely make the UI display the `SiteSelector` component and disrupt the user experience.

This is a rare issue in Calypso Blue, but it happens frequently in Calypso Green. This happens because we are not persisting the state of the app on the client-side as we do on Calypso Blue.

The demo section showcases an example of the issue in both environments.

#### Solution

Make persisting the state of the app default on all environments.

#### Changes proposed in this Pull Request

* Remove the `persist-redux` feature from all config files and make it default on every environment.

#### Testing instructions

Prerequisite: Jetpack site with a broken connection.

**The issue**
* Visit `cloud.jetpack.com/backup/:site` replacing `:site` with your Jetpack site that has a broken connection.
* Verify that you see the Site Selector component (re-visit the page if you don't see the Site Selector).

**The solution**
* Run this PR (both Calypsos).
* On Calypso Green, visit `/backup/:site` with the same site you tried before.
* Hard refresh the page.
* Verify that you see the Backup section (some of the times it might fail and you will see the `SiteSelector`).
* On Calypso Blue, visit `/backup/:site` with the same site you tried before.
* Hard refresh the page.
* Verify that you see the Backup section (some of the times it might fail and you will see the `SiteSelector`).

Fixes 1164141197617539-as-1199202377313720

#### Calypso Blue
Kapture 2020-11-24 at 12 29 55

#### Calypso Green
Kapture 2020-11-24 at 12 10 49

### CODE REVIEW

**Comment by rcanepa** [comment]
@jsnajdr I added you because I would like to have one or two reviews from Calypso's team. 

I couldn't find a proper label for this PR, so please feel free to add any label you think would be suitable.

**Comment by mattgawarecki** [comment]
The fix works for me! However, I'm curious how this will affect people who have lots of sites attached to their account. Will it cause page loads to take significantly longer, I wonder?

**Comment by jsnajdr** [comment]
Hi @rcanepa :wave: This change is very undesired from a performance point of view. For users with many sites, it will significantly slow down Calypso load. That group includes virtually all a12s, who commonly have hundreds of sites. For me personally, testing the PR on calypso.live, it introduces more than 5s delay.

> If the site has a broken connection, that request will fail.

It seems that it's the `/sites/:site` REST endpoint that really needs fixing here. Why does it fail when requesting a single site, and returns valid information happily when requesting the whole list? It shouldn't matter which way we're requesting the site.

When the site with a broken connection is part of the list, what does the record look like? Is it missing any significant information?

**Comment by rcanepa** [comment]
> Hi @rcanepa 👋 This change is very undesired from a performance point of view. For users with many sites, it will significantly slow down Calypso load. That group includes virtually all a12s, who commonly have hundreds of sites. For me personally, testing the PR on calypso.live, it introduces more than 5s delay.

I didn't notice a performance penalty when I tested it, but I was definitely concerned about it. 

What comes now are comments and questions meant to improve my understanding of the app and not trying to convince you that this change is good for the app.

In my limited experience, this is an edge case for Calypso Blue users. Normally, the app's state has the list of sites at the time the runtime hits the `siteSelection` controller, so it almost never has to do the request. So, your comment makes me wonder if there is a client-side cache that stores the list of sites and that's why on Calypso Blue this bug almost never happens. Could we have a client-site cache that works only on Calypso Blue and not on Calypso Green?

I understand that the `requestSites` action is always called regardless of the change I made because it's the source of data of the `SiteSelector` component which, I believe, is almost always present.

Additionally, I have seen that we tend to use the `siteSelection` controller along with the `SiteSelector` component. So, if the `siteSelection` controller doesn't find the selected site, the app will display the `SiteSelector` component to prompt the user to select a site from the list. So, again, the list of sites must be present.

In the end, I totally agree that we shouldn't add this change if it's going to slow down the app or made it request a piece of data that was already requested (duplicated requests).

> It seems that it's the `/sites/:site` REST endpoint that really needs fixing here. Why does it fail when requesting a single site, and returns valid information happily when requesting the whole list? It shouldn't matter which way we're requesting the site.

This is a really good point. I'll definitely explore the backend to understand why the endpoint behaves like this and if we can make it better. What I don't like about this option though, is that I would potentially break clients that expected the current behavior. Updating how Calypso works would be a much more controlled change.

Thank you for your feedback @jsnajdr 🙇

**Comment by rcanepa** [comment]
> When the site with a broken connection is part of the list, what does the record look like? Is it missing any significant information?

I don't think anything is missing when the connection is broken. However, the information we get about a site from `/sites` is less than what we get from `/site/:site`. Specifically, `/site/:site` returns these additional properties:
```
[
  "is_following",
  "jetpack_connection",
  "locale",
  "logo",
  "meta",
  "post_count",
  "quota",
  "subscribers_count"
]
```

**Comment by ChaosExAnima** [comment]
> In my limited experience, this is an edge case for Calypso Blue users. Normally, the app's state has the list of sites at the time the runtime hits the `siteSelection` controller, so it almost never has to do the request. So, your comment makes me wonder if there is a client-side cache that stores the list of sites and that's why on Calypso Blue this bug almost never happens. Could we have a client-site cache that works only on Calypso Blue and not on Calypso Green?

One thing I'm noticing @rcanepa - if you check the source code, Calypso Blue has user info already loaded by the server into the `currentUser` variable. Calypso Green does not. I wonder if there's some logic on the server side that's different?

**Comment by jsnajdr** [comment]
> In my limited experience, this is an edge case for Calypso Blue users.

Yes, this is mostly true. The issue affects mainly a12s, and only if the sites list is not present or not up to date in the local storage cache (more on that below). Despite that, I remember that it was a very visible issue before we fixed it in early 2018 (p9oQ9f-11-p2 is a good P2 to start with if you want to learn more), and if we de-optimized the sites load now, it's very likely that reports would start popping up quickly. ("Howdy, my Calypso sometimes takes very long to load, with a 5+ seconds delay. Happens randomly, most of the time it's fine. Any idea what's going on? It didn't start until recently.")

If we can make it work by fixing the REST API, that would be vastly preferrable.

> So, your comment makes me wonder if there is a client-side cache that stores the list of sites and that's why on Calypso Blue this bug almost never happens.

Yes, there is a client-side cache 🙂 The `/me/sites` response is stored in Redux store, and the store's state is regularly persisted to IndexedDB. You can see the database in your devtools, in the IndexedDB section, called `calypso_store`.

When you open Calypso in a new tab, the data from IndexedDB is loaded into Redux on boot. So you have a sites list from the very beginning, it just might be stale. The `/me/sites` or `/sites/:site` REST request then merely refreshes the site info, instead of loading it from scratch. `siteSelection` uses the cached data and is able to do the initial render quickly.

> Could we have a client-site cache that works only on Calypso Blue and not on Calypso Green?

This is also true, and I didn't know that Calypso Green (aka Jetpack Cloud) has the cache disabled. There is a `persist-redux` feature flag that enables it, and it's enabled only in Calypso Blue environments. I don't know why is that -- it's a rather harmless thing that could be enabled everywhere. It looks more like a mistake by omission rather than something intentional. Maybe @tyxla knows?

> I understand that the requestSites action is always called regardless of the change I made because it's the source of data of the SiteSelector component which, I believe, is almost always present.

Yes, the `/me/sites` request is always done on boot, in parallel with the site-specific and much faster `/sites/:site` one. The difference is that waiting for `/me/sites` is not on the critical path and doesn't block the initial render. We only need to wait for the fast `/sites/:site`. The `/me/sites` response arrives a bit later, when the user is already looking at the desired UI (e.g., Stats) instead of a gray placeholder.

> What I don't like about this option though, is that I would potentially break clients that expected the current behavior.

Given that Calypso is the primary consumer of this REST API by a huge margin (challenged maybe by the mobile apps), I think this risk is minimal. If we want to mitigate it completely, we can add a query param to the request: `/sites/:site?include_broken_jetpacks=1`.

> if you check the source code, Calypso Blue has user info already loaded by the server into the currentUser variable. Calypso Green does not. I wonder if there's some logic on the server side that's different?

@ChaosExAnima Yes, there is a feature called "user bootstrap", controlled by the `wpcom-user-bootstrap` feature flag. It's enabled only in production environment, because the server needs to know certain secret keys to make it work. If enabled, the server itself, when asked for the `/` HTML resource, will make a rest request to `/me`, with the auth cookies from the `/` request, and return the user info embedded in the HTML. That saves the client at least one XHR roundtrip on the critical path.

This, again, is not enabled in Jetpack Cloud. Probably because the current implementation of user bootstrap can work only with wordpress.com auth cookies, not with OAuth tokens that Jetpack Cloud uses. It shouldn't be hard to add, however, if we decided we want it in Jetpack Cloud, too. Again, maybe @tyxla knows more.

> I don't think anything is missing when the connection is broken. However, the information we get about a site from /sites is less than what we get from /site/:site.

@rcanepa I'd like to reproduce these REST responses, but I don't know how. What's the best way to simulate a broken connection on a Jetpack testing site? When I simply disabled the Jetpack plugin, I saw a different behavior: the site completely disappeared from the `/me/sites` list, and `/sites/:site` started returning a "API calls to this blog have been disabled" error. Apparently, disabling the plugin disconnected the site gracefully. I need some recipe how to break the connection better 🙂

**Comment by tyxla** [comment]
> There is a persist-redux feature flag that enables it, and it's enabled only in Calypso Blue environments. I don't know why is that -- it's a rather harmless thing that could be enabled everywhere. It looks more like a mistake by omission rather than something intentional. Maybe @tyxla knows?

Yep, it seems like it's a mistake by omission indeed. IIRC, enabled feature flags in the Jetpack Cloud environments were hand-picked at the time, so it's easy to miss something like `persist-redux`. I also agree that it should be pretty harmless to enable in Jetpack Cloud, and I'd advise doing that - it can be pretty beneficial.

> This, again, is not enabled in Jetpack Cloud. Probably because the current implementation of user bootstrap can work only with wordpress.com auth cookies, not with OAuth tokens that Jetpack Cloud uses. It shouldn't be hard to add, however, if we decided we want it in Jetpack Cloud, too. Again, maybe @tyxla knows more.

Back when we were creating the environments, we hadn't explored what mechanisms we would be using for auth at all. That explains why we weren't using the user bootstrap. Later, I wasn't involved in the conversations and ongoing work to implement OAuth there. Maybe the idea was to decouple more from the existing auth mechanisms and make it more a separate app, or to make it more sustainable for the future where 3rd party domain cookies would be a no-go, but I wouldn't know for sure, as I wasn't involved in that work.

> > Hi @rcanepa 👋 This change is very undesired from a performance point of view. For users with many sites, it will significantly slow down Calypso load. That group includes virtually all a12s, who commonly have hundreds of sites. For me personally, testing the PR on calypso.live, it introduces more than 5s delay.

> I didn't notice a performance penalty when I tested it, but I was definitely concerned about it.

I'd like to second Jarda's comment here - as someone who has over 500 sites already, I can confirm such a change can be devastating for performance. On the contrary, requesting a single site vs all sites made gutenframe 50% faster for users with hundreds of sites (see p4TIVU-9uP-p2 for context). So I'd advise against requesting all sites when possible. Instead, is it possible to call a `requestSites` at the controller level or `QuerySites` at the component level where needed?

**Comment by mattgawarecki** [comment]
@rcanepa @ChaosExAnima @tyxla @jsnajdr Just stopping in again to express how amazing the discussion on this PR has been. I'm learning so much!

**Comment by rcanepa** [comment]
> @rcanepa @ChaosExAnima @tyxla @jsnajdr Just stopping in again to express how amazing the discussion on this PR has been. I'm learning so much!

This x 100!

> The issue affects mainly a12s, and only if the sites list is not present or not up to date in the local storage cache (more on that below). Despite that, I remember that it was a very visible issue before we fixed it in early 2018 (p9oQ9f-11-p2 is a good P2 to start with if you want to learn more), and if we de-optimized the sites load now, it's very likely that reports would start popping up quickly. ("Howdy, my Calypso sometimes takes very long to load, with a 5+ seconds delay. Happens randomly, most of the time it's fine. Any idea what's going on? It didn't start until recently.")

Thanks for providing a link to that P2. It was an awesome starting point to understand why things evolved into what they are now. I highly recommend p9oQ9f-Q-p2 to anyone interested in reading a performance breakdown of the loading of `/stats/day/:site`. It touches a lot of the points that were discussed here.

> Yes, there is a client-side cache 🙂 The /me/sites response is stored in Redux store, and the store's state is regularly persisted to IndexedDB. You can see the database in your devtools, in the IndexedDB section, called calypso_store.

I suspected this would be the case because when I refreshed the page, things loaded very fast.

> When you open Calypso in a new tab, the data from IndexedDB is loaded into Redux on boot. So you have a sites list from the very beginning, it just might be stale. The /me/sites or /sites/:site REST request then merely refreshes the site info, instead of loading it from scratch. siteSelection uses the cached data and is able to do the initial render quickly.

This is a really interesting feature of the app and we definitely should make use of it in Calypso Green.

> This is also true, and I didn't know that Calypso Green (aka Jetpack Cloud) has the cache disabled. There is a persist-redux feature flag that enables it, and it's enabled only in Calypso Blue environments. I don't know why is that -- it's a rather harmless thing that could be enabled everywhere. It looks more like a mistake by omission rather than something intentional.

I'm going to test enabling the flag in Calypso Green. Maybe it will make the issue almost disappear just like it happens in Calypso Blue.

> Yes, the /me/sites request is always done on boot, in parallel with the site-specific and much faster /sites/:site one. The difference is that waiting for /me/sites is not on the critical path and doesn't block the initial render. We only need to wait for the fast /sites/:site. The /me/sites response arrives a bit later, when the user is already looking at the desired UI (e.g., Stats) instead of a gray placeholder.

This is a really good point and why we shouldn't proceed with the change I made. 

> Given that Calypso is the primary consumer of this REST API by a huge margin (challenged maybe by the mobile apps), I think this risk is minimal. If we want to mitigate it completely, we can add a query param to the request: /sites/:site?include_broken_jetpacks=1.

I'm glad to read this. I didn't know that Calypso is the primary consumer. I'll definitely spend more time exploring this option.

> @rcanepa I'd like to reproduce these REST responses, but I don't know how. What's the best way to simulate a broken connection on a Jetpack testing site? When I simply disabled the Jetpack plugin, I saw a different behavior: the site completely disappeared from the /me/sites list, and /sites/:site started returning a "API calls to this blog have been disabled" error. Apparently, disabling the plugin disconnected the site gracefully. I need some recipe how to break the connection better 🙂

I know we can simulate a broken connection using [Jetpack Debug tools](https://github.com/Automattic/jetpack/tree/master/packages/debug-helper) but these tools are only available to the site that comes with the Jetpack repository. I think I read that these tools are going to be available to JN sites soon. Having said that, there is another way that I have been using, but you need a deleted JN site that is still connected to your WordPress.com. In that case, the site will appear in your site list and WordPress.com will think that the connection is broken. If you have a site like that, and you visit Calypso Green, you will immediately see the problem. I just learned that if you clean the client-side cache in Calypso Blue, you will see the problem as well.

**Comment by rcanepa** [comment]
> Yep, it seems like it's a mistake by omission indeed. IIRC, enabled feature flags in the Jetpack Cloud environments were hand-picked at the time, so it's easy to miss something like persist-redux. I also agree that it should be pretty harmless to enable in Jetpack Cloud, and I'd advise doing that - it can be pretty beneficial.

Cool. I'll definitely try adding this feature to Calypso Green. Hopefully, it will fix the issue for the second time a user lands in the app.

> I'd like to second Jarda's comment here - as someone who has over 500 sites already, I can confirm such a change can be devastating for performance. On the contrary, requesting a single site vs all sites made gutenframe 50% faster for users with hundreds of sites (see p4TIVU-9uP-p2 for context). 

Thanks for that P2. It's absolutely clear to me that we shouldn't request all sites to fix this issue.

> Instead, is it possible to call a requestSites at the controller level or QuerySites at the component level where needed?

I'm not sure it's possible. If you go to `cloud.jetpack.com/backup/:site`, you will have to go through the `siteSelection` controller (and many others) before you even have a chance to get to the Backup section. But I'll explore this idea, maybe there is a component higher in the tree that we could leverage for that.

**Comment by rcanepa** [comment]
I tested enabling the `persist-redux` feature and it definitely makes the app load faster. The issue seems to appear less frequently as well.

I pushed a rebased version of the PR with that change and reverting the change I made to the `siteSelection` controller.

File: config/jetpack-cloud-development.json [hunk 1]
```diff
@@ -44,6 +44,7 @@
 		"layout/query-selected-editor": false,
 		"layout/support-article-dialog": false,
 		"oauth": true,
+		"persist-redux": true,
```
Comment (jsnajdr) [inline config/jetpack-cloud-development.json]: I'd consider removing the `persist-redux` feature flag completely, and assume it's always `true`.

The flag was probably added at a time when Redux persistence was new and risky. But now it's been enabled for many years, and became boring 🙂

Comment (rcanepa) [reply config/jetpack-cloud-development.json]: That's a great idea. There is no need to have it if it's enabled in every environment. I'll clean that up.

**Review by monsieur-z — APPROVED** [review]
Great PR, lots of insights! It tests well, I haven't seen the site selection page. I also verified in my editor that I couldn't find any instance of `'persist-redux'` anymore.

**Comment by rcanepa** [comment]
Hey again @jsnajdr, I made the `persist-redux` feature default for every env. And since this is as far as this PR is going to get because the rest of the work will happen in the backend, I would really appreciate another review from yours.

File: client/state/test/initial-state.js [hunk 2]
```diff
@@ -44,7 +44,7 @@ jest.mock( 'config', () => {
 
 	const config = () => 'development';
 
-	config.isEnabled = jest.fn( ( key ) => key === 'persist-redux' && persistReduxEnabled );
+	config.isEnabled = jest.fn( () => persistReduxEnabled );
```
Comment (jsnajdr) [inline client/state/test/initial-state.js]: In this test, the entire mock for `config`, and all the calls to `{en,dis}ablePersistRedux`, can be safely removed. The persistence code doesn't depend on any `config` flag any more.

**Review by jsnajdr — APPROVED** [review]
A lot more code can be removed from the`initial-state` test. Otherwise, the change looks good :+1: Thanks for working on this.

File: client/state/test/initial-state.js (hunk 2 above)
Comment (rcanepa) [reply client/state/test/initial-state.js]: Oh, you're totally right about this!

**Pull request merged**